kubeadm alpha certs generate-csr

wallrj opened the pull request:
/kind feature

This is designed to supplant the `kubeadm init phase certs ... --csr-only` sub-command which was found to be difficult to use because it also runs all the validation steps that are run as part of `kubeadm init`.

Deprecated `--csr-only` and `--csr-dir` flags 

Ref: https://github.com/kubernetes/kubeadm/issues/2163
 
```release-note
kubeadm: deprecate the "--csr-only" and "--csr-dir" flags of the "kubeadm init phase certs" subcommands. Please use "kubeadm alpha certs generate-csr" instead. This new command allows you to generate new private keys and certificate signing requests for all the control-plane components, so that the certificates can be signed by an external CA.
```

neolit123 commented:
/ok-to-test

neolit123 commented:
/assign @yagonobre @rosti

rosti commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +87,157 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
+		RunE: func(cmd *cobra.Command, args []string) error {
+			config, err := flagValues.toConfig()
+			if err != nil {
+				return errors.Wrap(err, "failed to generate config")
```
No need for a `Wrap` here. Just return the error as is.
Besides adding an additional message, `Wrap` would also create a stack trace and attach it to the resulting error. Chances are, that this has already been done by `toConfig` or some other function that's called by it.
In general, `Wrap` is placed when we call functions that are external to the kubeadm code base and are not generic `utils` types of functions.

[1]

rosti commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +87,157 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
+	certDir := config.kubeadmConfig.CertificatesDir
+	certTree, err := certsphase.GetDefaultCertList().AsMap().CertTree()
+	if err != nil {
+		return errors.Wrap(err, "failed to create cert tree")
```
[1]

rosti commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +87,157 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
+			name := cert.BaseName
+			cfg, err := cert.GetConfig(config.kubeadmConfig)
+			if err != nil {
+				return errors.Wrap(err, "failed to get config")
```
[1]

rosti commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +87,157 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
+			}
+			csr, key, err := pkiutil.NewCSRAndKey(cfg)
+			if err != nil {
+				return errors.Wrap(err, "failed to create csr or key")
```
[1]

rosti commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +87,157 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
+			}
+			err = pkiutil.WriteKey(certDir, name, key)
+			if err != nil {
+				return errors.Wrap(err, "failed to write key")
```
[1]

rosti commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +87,157 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
+			}
+			err = pkiutil.WriteCSR(certDir, name, csr)
+			if err != nil {
+				return errors.Wrap(err, "failed to write CSR")
```
[1]

rosti commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +87,157 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
+	}
+	klog.V(1).Infoln("[generate-csr] Generating private keys and CSRs for kubeconfig files")
+	if err := kubeconfigphase.CreateJoinControlPlaneKubeConfigAndCSRFiles(config.kubeConfigDir, config.kubeadmConfig); err != nil {
+		return errors.Wrap(err, "failed to create kubeconfig files")
```
[1]

rosti commented on cmd/kubeadm/app/cmd/alpha/certs_test.go:
```diff
@@ -27,6 +29,13 @@ import (
 	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
```
Better put a new line after this one. We separate `k8s.io` from other packages.

rosti commented on cmd/kubeadm/app/cmd/alpha/certs_test.go:
```diff
@@ -285,3 +294,212 @@ func TestRenewUsingCSR(t *testing.T) {
+kubernetesVersion: v1.19.0
+`
+
+func TestGenCSR(t *testing.T) {
```
This looks more like an integration test to me. Possibly, we need to refactor and put it in `cmd/kubeadm/test/cmd`.

rosti commented on cmd/kubeadm/app/util/kubeconfig/builder/builder.go:
```diff
@@ -0,0 +1,133 @@
+package builder
```
Overall, I like the idea here. I think that this can greatly simplify and make the code more understandable.
But I think, that this is outside of the scope of this PR.
Possibly spit it up as a separate PR with the idea of cleaning up the contents and interface of `kubeconfig.go`?

rosti reviewed (commented):
Thanks @wallrj !
It's nice for a first kubeadm PR that is this large!

yagonobre commented:
Thanks for work on it @wallrj you can fix the bazel test with `./hack/update-bazel.sh`

yagonobre commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +87,157 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
 }
 
+// genCSRConfig is the configuration required by the gencsr command
+type genCSRConfig struct {
```
+1

yagonobre reviewed (commented):
Thanks for working on it, in general LGTM

wallrj reviewed (commented):
I've removed the kubeconfig builder package for now and I'll create a separate PR for that.
I'll look into moving the integration test along side other such tests.

wallrj commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +87,157 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
+		Use:     "generate-csr",
+		Aliases: []string{"generate-keys-and-certificate-signing-requests"},
+		Short:   "Generate keys and certificate signing requests",
+		Long: cmdutil.LongDesc(`
```
Done.

wallrj reviewed (commented):
Moved description and examples to global vars for consistency with other commands.

wallrj commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +87,157 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
+	}
+}
+
+func (o *genCSRFlags) addFlagSet(flagSet *pflag.FlagSet) {
```
That would make it more difficult to unit test. I'd need to create a cobra command in the unit test and have a way to make it parse the flags.

rosti commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +129,111 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
+// else use dynamically generated default config.
+// This configuration contains the DNS names and IP addresses which
+// are encoded in the control-plane CSRs.
+func (o *genCSRConfig) finalize() (err error) {
```
`finalize` is a bit confusing name here. Sounds like a Java finalize method. I'd rather change it to `load`, `prepare` or something on those lines.

rosti commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +129,111 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
+// This configuration contains the DNS names and IP addresses which
+// are encoded in the control-plane CSRs.
+func (o *genCSRConfig) finalize() (err error) {
+	o.kubeadmConfig, err = configutil.LoadOrDefaultInitConfiguration(
```
Is this command supposed to be invoked without a config file? If that's the case, then we may have to introduce the rest of the config command line flags here.

neolit123 replied:
@wallrj should correct me here, but i think if the user is setting custom SANs under ClusterConfiguration, then the new command should be aware of the --config.

wallrj replied:
I'd need to add some of the other flags that are currently supported by `kubectl init phase certs all`

```
_output/bin/kubeadm init phase certs all --help
Generate all certificates

Usage:
  kubeadm init phase certs all [flags]

Flags:
      --apiserver-advertise-address string   The IP address the API Server will advertise it's listening on. If not set the default network interface will be used.
      --apiserver-cert-extra-sans strings    Optional extra Subject Alternative Names (SANs) to use for the API Server serving certificate. Can be both IP addresses and DNS names.
      --cert-dir string                      The path where to save and store the certificates. (default "/etc/kubernetes/pki")
      --config string                        Path to a kubeadm configuration file.
      --control-plane-endpoint string        Specify a stable IP address or DNS name for the control plane.
  -h, --help                                 help for all
      --kubernetes-version string            Choose a specific Kubernetes version for the control plane. (default "stable-1")
      --service-cidr string                  Use alternative range of IP address for service VIPs. (default "10.96.0.0/12")
      --service-dns-domain string            Use alternative domain for services, e.g. "myorg.internal". (default "cluster.local")


```

* `--api-server-cert-extra-sans`  https://github.com/kubernetes/kubernetes/blob/c5941e283fd7e37e953832d6743e7ea3b556ab7a/cmd/kubeadm/app/cmd/init.go#L248-L251 

Or I could require that the user supplies a kubeadm config file and puts all those settings in there.

rosti replied:
I think, that we can start simple and require a config file for now. If many users start complaining we can add these extra options.

wallrj replied:
As discussed  in office hours meeting last week, I have not added the extra command line options.
The `--config` flag is optional. If  not supplied, the Kubeadm config will be dynamically generated based on local machine settings.

rosti reviewed (commented):
Thanks @wallrj !
Looks like we are on the final stretch here.

neolit123 commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -68,6 +74,47 @@ var (
+	│   ├── front-proxy-client.csr
+	│   └── front-proxy-client.key
+	├── scheduler.conf
+	└── scheduler.conf.csr
```
i'd prefer if we don't hardcode the list list of CSR/keys as part of the example, because if we end up changing something this hardcoded list will drift.

IMO the first part is suffcient:
```
# The following command will generate keys and CSRs for all control-plane certificates and kubeconfig files:
kubeadm alpha certs generate-csr --kubeconfig-dir /tmp/etc-k8s --cert-dir /tmp/etc-k8s/pki
```

neolit123 commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +129,111 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
+		return err
+	}
+	for caCertificate, leafCertificates := range certTree {
+		klog.V(1).Infoln("[generate-csr] Generating private keys and CSRs for leaf certificates of", caCertificate.Name)
```
should be `Infof(....%q, caCertificate.Name`), in kubeadm we mostly use Infof (or Printf from stdlib), instead of the multi-arg concat that Infoln (Println) has.

wallrj replied:
Ah, I see that other commands use `fmt.Printf` for progress messages. I've removed all that logged output. It was unnecessary.

neolit123 commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +129,111 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
+	for caCertificate, leafCertificates := range certTree {
+		klog.V(1).Infoln("[generate-csr] Generating private keys and CSRs for leaf certificates of", caCertificate.Name)
+		for _, cert := range leafCertificates {
+			klog.V(1).Infoln("[generate-csr] :: ", cert.Name)
```
ditto.

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -112,63 +126,116 @@ func createKubeConfigFiles(outDir string, cfg *kubeadmapi.InitConfiguration, kub
+			return err
+		}
+
+		// writes the kubeconfig to disk if it not exists
```
> ...if it does not exist

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -112,63 +126,116 @@ func createKubeConfigFiles(outDir string, cfg *kubeadmapi.InitConfiguration, kub
+		}
+		// Write CSR to disk
+		if err := pkiutil.WriteCSR(kubeConfigDir, kubeConfigFileName, csr); err != nil {
+			return errors.Wrap(err, "failed to write CSR file")
```
returning the error here without Wrap is fine as the underlying errors are descriptive:
https://github.com/kubernetes/kubernetes/blob/4e79344501bfc5d3bc48caa12c41b9246d90850d/cmd/kubeadm/app/util/pkiutil/pki_helpers.go#L166-L181

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -112,63 +126,116 @@ func createKubeConfigFiles(outDir string, cfg *kubeadmapi.InitConfiguration, kub
 }
 
+// getKubeConfigSpecs returns all KubeConfigSpecs actualized to the context of the current InitConfiguration
+// NB. this methods holds the information about how kubeadm creates kubeconfig files.
```
we seem to have the same NB line here and for getKubeConfigSpecsBase.
should we only have it for one of the functions?

neolit123 commented:
> Fixes kubernetes/kubeadm#2163

`s/Fixes/ref`

i think we should keep that issue open to track the removal of the deprecated flags.

neolit123 commented:
/priority important-longterm
/milestone v1.19

wallrj committed:
Deprecate --csr-only and --csr-dir flags

Signed-off-by: Richard Wall <richard.wall@jetstack.io>

neolit123 commented:
continuing from the discussion from today at the kubeadm office hours.

A) "--config"

https://github.com/kubernetes/kubernetes/pull/92183#discussion_r447002061

we shouldn't bind the flags in more places, a long term plan for kubeadm is to start removing some of these flags that are covered by in the API / config. we should 1) make --config the source of the options for the new command 2) if no --config is added assume defaults for the node where the command is called.

B) CLI integration vs unit tests.

we should just unit test runGenCSR(). the unit test should be called TestRunGenCSR().

neolit123 commented:
hello @wallrj 
@rosti is on PTO this week so i'd have to finish the reviews here.

neolit123 commented on cmd/kubeadm/app/cmd/options/certs.go:
```diff
@@ -26,9 +26,11 @@ func AddCertificateDirFlag(fs *pflag.FlagSet, certsDir *string) {
 // AddCSRFlag adds the --csr-only flag to the given flagset
 func AddCSRFlag(fs *pflag.FlagSet, csr *bool) {
 	fs.BoolVar(csr, CSROnly, *csr, "Create CSRs instead of generating certificates")
+	fs.MarkDeprecated(CSROnly, "This flag will be removed in a future version. Please use kubeadm alpha certs generate-csr instead.")
```
so i was going to recommend to include the deprecation message as part of the release note under "User facing change" in the PR description, but i realized the flags are used in a couple of places:
- kubeadm init phase certs
- kubeadm alpha certs renew

so the question here is - should we deprecate these flags in all places? (cc @fabriziopandini)

unlike "phase certs" where CSRs are less useful, for "renew" one might want to not renew right away and only generate the CSRs. i guess at this point we will be pointing users towards the new command that handles CSRs specifically and "renew" will be all about in-place renewal using the CA kubeadm looks for on disk? are my UX assumptions correct here?

in any case, we need to mention for which commands we are deprecating these flags in the release note.

wallrj replied:
Ok. I've deprecated those flags for only `kubectl init phase certs ...` , 
not for `kubectl alpha certs renew --csr-only --csr-dir`.

wallrj commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -209,6 +184,42 @@ func buildKubeConfigFromSpec(spec *kubeConfigSpec, clustername string) (*clientc
+		emptyClientCert,
+	), clientCSR, nil
+}
+
```
Note to self: Move this to create.

wallrj commented on cmd/kubeadm/app/util/pkiutil/pki_helpers.go:
```diff
@@ -288,13 +288,23 @@ func TryLoadKeyFromDisk(pkiPath, name string) (crypto.Signer, error) {
+func TryLoadCSRAndKeyFromDisk(pkiPath, name string) (*x509.CertificateRequest, crypto.Signer, error) {
+	csr, err := TryLoadCSRFromDisk(pkiPath, name)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't load csr file")
```
Not to self: Test this.

wallrj commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +102,93 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
 	return cmd
 }
 
+// runGenCSR contains the logic of the generate-csr sub-command.
+func runGenCSR(config *genCSRConfig) error {
+	leafCertificates, err := certsphase.LeafCertificates(certsphase.GetDefaultCertList())
+	if err != nil {
+		return err
+	}
+	if err := certsphase.CreateKeyAndCSRFiles(config.kubeadmConfig, leafCertificates); err != nil {
+		return err
+	}
+
+	kubeConfigs, err := kubeconfigphase.GetDefaultKubeConfigs(config.kubeadmConfig)
+	if err != nil {
+		return err
+	}
+	if err := kubeconfigphase.CreateKubeConfigAndCSRFiles(config.kubeConfigDir, config.kubeadmConfig, kubeConfigs); err != nil {
+		return err
+	}
+	return nil
+}
```
* moved all the certificate CSR code into `cmd/kubeadm/app/phases/certs` with unit tests
 * moved all the kubeconfig CSR code into `cmd/kubeadm/app/phases/kubeconfig` with unit tests
 * Tried to bring more consistency between these two packages, with both now having a  public `GetDefaultXXX` function. 
 * Both have a new `CreateXXXAndCSRFiles` function which accepts a list of things.
 * This function has a (slow) unit test for the happy path only, but each of the packages has more complete unit tests for error cases.

wallrj commented on cmd/kubeadm/app/cmd/alpha/certs_test.go:
```diff
@@ -285,3 +293,188 @@ func TestRenewUsingCSR(t *testing.T) {
+		assert.NoError(t, err, "failed to load kubeconfig CSR file: %s", name)
+	}
+
+}
```
This is a test for the happy path only. Error cases in the creation of cert CSR and kubeconfig CSR files are tested in the respective packages.

wallrj commented on cmd/kubeadm/app/cmd/phases/init/certs.go:
```diff
@@ -66,7 +66,9 @@ func NewCertsPhase() workflow.Phase {
+	set.MarkDeprecated(options.CSROnly, "This flag will be removed in a future version. Please use kubeadm alpha certs generate-csr instead.")
 	options.AddCSRDirFlag(set, &csrDir)
+	set.MarkDeprecated(options.CSRDir, "This flag will be removed in a future version. Please use kubeadm alpha certs generate-csr instead.")
 	return set
```
Deprecated these flags for the `kubeadm init phase certs` sub-command only. As suggested by @neolit123

wallrj commented on cmd/kubeadm/app/phases/certs/certlist.go:
```diff
@@ -204,6 +207,22 @@ func (c Certificates) AsMap() CertificateMap {
+		}
+	}
+	return nil
+}
```
This allows me to unit test the  behaviour of Certificates list iteration with simple visitor functions.
Avoids having to have slow unit tests that create lists of keys and CSR files.

wallrj commented on cmd/kubeadm/app/phases/certs/certlist.go:
```diff
@@ -397,3 +416,58 @@ func setCommonNameToNodeName() configMutatorsFunc {
+		out = append(out, leafCertificates...)
+	}
+	return out, nil
+}
```
I moved this to a separate function where it can be unit tested separately from the complex and slow key and CSR generation.

wallrj commented on cmd/kubeadm/app/phases/certs/certlist.go:
```diff
@@ -397,3 +416,58 @@ func setCommonNameToNodeName() configMutatorsFunc {
+		return err
+	}
+	return nil
+}
```
This where I can test the creation of a single key and CSR file and *some* of the error cases.
This could be made more generic  and reusable in future. See https://github.com/kubernetes/kubernetes/compare/master...wallrj:kubeconfig-builder

wallrj commented on cmd/kubeadm/app/phases/certs/certlist.go:
```diff
@@ -397,3 +416,58 @@ func setCommonNameToNodeName() configMutatorsFunc {
+// adjacent CSR files.
+func CreateKeyAndCSRFiles(config *kubeadmapi.InitConfiguration, certificates Certificates) error {
+	return certificates.visit((&keyAndCSRCreator{kubeadmConfig: config}).create)
+}
```
This then becomes a simple wrapper which only needs a unit test for the happy path (see below)

wallrj commented on cmd/kubeadm/app/phases/certs/certlist_test.go:
```diff
@@ -209,6 +214,312 @@ func TestCreateCertificateChain(t *testing.T) {
+
+			if test.assertion == nil {
+				test.assertion = func(t *testing.T, tc *testCase) {
+					assert.Len(t, tc.result, 3)
```
I should add more assertions here, to check that only non-CA certificates are returned, for example.

wallrj commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -209,6 +171,16 @@ func buildKubeConfigFromSpec(spec *kubeConfigSpec, clustername string) (*clientc
+		},
+	}
+}
+
```
I split this simple snippet into a function so that I can get the same config in the new function below.

wallrj commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -386,3 +358,141 @@ func ValidateKubeconfigsForExternalCA(outDir string, cfg *kubeadmapi.InitConfigu
 }
+
+// KubeConfigs is a mapping between kubeconfig file name and spec
+type KubeConfigs map[string]*kubeConfigSpec
```
There was already a `map[string]*kubeConfigSpec` in the original code, so I added an alias for it and attached a `visit` method.
I'd rather this be a slice, consistent with `certsphase.Certificates` but that would have required more changes in an already large PR.

wallrj commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -386,3 +358,141 @@ func ValidateKubeconfigsForExternalCA(outDir string, cfg *kubeadmapi.InitConfigu
+	}
+	kubeConfigPath := filepath.Join(o.kubeConfigDir, name)
+	if _, err := os.Stat(kubeConfigPath); err == nil {
+		return fmt.Errorf("%w: kube config: %s", os.ErrExist, kubeConfigPath)
```
`os.ErrExist` seems like a natural error to return here, but I considered adding a package level `errExist` type instead.

wallrj commented on cmd/kubeadm/app/util/pkiutil/pki_helpers.go:
```diff
@@ -334,6 +332,18 @@ func TryLoadPrivatePublicKeyFromDisk(pkiPath, name string) (*rsa.PrivateKey, *rs
+	}
+
+	return csr, nil
+}
```
This function is just a wrapper around ` CertificateRequestFromFile` with added filename generation so I didn't add a unit test.
And it's only used in tests anyway.

wallrj reviewed (commented):
@neolit123 I've added quite a lot of new unit tests, which has made this a much larger PR, sorry. I've added a few explanatory comments below. I hope these make sense.

neolit123 commented:
```

staticcheck expand_less | 6m0s
-- | --
Errors from staticcheck: cmd/kubeadm/app/phases/certs/certlist_test.go:414:3: field tmpDir is unused (U1000)  Please review the above warnings. You can test via:   hack/verify-staticcheck.sh <failing package>


```
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/92183/pull-kubernetes-verify/1280503024111849472

neolit123 commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -24,17 +24,21 @@ import (
+	"github.com/spf13/pflag"
 
 	"k8s.io/apimachinery/pkg/util/duration"
+
```
new line is not needed here.

neolit123 commented on cmd/kubeadm/app/cmd/alpha/certs_test.go:
```diff
@@ -27,6 +27,14 @@ import (
+	"github.com/stretchr/testify/require"
+
+	"k8s.io/client-go/tools/clientcmd"
+
```
new line is not needed here.

neolit123 commented on cmd/kubeadm/app/cmd/alpha/certs_test.go:
```diff
@@ -285,3 +293,188 @@ func TestRenewUsingCSR(t *testing.T) {
+
+	t.Log("The command generates kubeconfig files in the configured --kubeconfig-dir")
+	for _, name := range expectedKubeConfigs {
+		t.Log("::", name)
```
we can omit all t.Log() calls from above.

neolit123 commented on cmd/kubeadm/app/phases/certs/certlist.go:
```diff
@@ -19,10 +19,13 @@ package certs
 	"github.com/pkg/errors"
 
 	certutil "k8s.io/client-go/util/cert"
+
```
empty line can be removed.

neolit123 commented on cmd/kubeadm/app/cmd/alpha/certs_test.go:
```diff
@@ -285,3 +293,188 @@ func TestRenewUsingCSR(t *testing.T) {
+		_, err = pkiutil.TryLoadCSRFromDisk(kubeConfigDir, name+".conf")
+		assert.NoError(t, err, "failed to load kubeconfig CSR file: %s", name)
+	}
+
```
empty line can be removed.

neolit123 commented on cmd/kubeadm/app/cmd/alpha/certs_test.go:
```diff
@@ -285,3 +293,188 @@ func TestRenewUsingCSR(t *testing.T) {
+	}
+
+	err := runGenCSR(&config)
+	require.NoError(t, err)
```
IMO some NoError() calls need annotations so that the person debugging knows where to find the problem in the test code.
`require.NoError(t, err, "expected runGenCSR to not fail")`

NoError does not print a stack trace, correct?
[1]

neolit123 commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -82,9 +103,108 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
+			err := config.load()
+			if err != nil {
+				return err
+			}
```
```
if err := config.load(); err != nil {
	return err
}
```

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -209,6 +263,52 @@ func buildKubeConfigFromSpec(spec *kubeConfigSpec, clustername string) (*clientc
+
+	encodedClientKey, err := keyutil.MarshalPrivateKeyToPEM(clientKey)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to marshal private key to PEM")
```
errors.Wrap

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -31,12 +31,12 @@ import (
 	certutil "k8s.io/client-go/util/cert"
 	"k8s.io/client-go/util/keyutil"
 	"k8s.io/klog/v2"
+
```
this empty line is not needed.

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -112,60 +126,105 @@ func createKubeConfigFiles(outDir string, cfg *kubeadmapi.InitConfiguration, kub
+// If kubeconfig files already exists, they are used only if evaluated equal; otherwise an error is returned.
+// This is a variant for use in CreateJoinControlPlaneKubeConfigAndCSRFiles which generates
+// CSR files instead of signed certificates.
+func createKubeConfigAndCSRFiles(kubeConfigDir string, cfg *kubeadmapi.InitConfiguration, kubeConfigFileNames ...string) error {
```
i wish we could have avoided some of these refactors to conform unit testing so close to feature freeze...

neolit123 commented on cmd/kubeadm/app/cmd/alpha/BUILD:
```diff
@@ -21,6 +21,7 @@ go_library(
         "//cmd/kubeadm/app/cmd/util:go_default_library",
```
this commit could have been split further:
- logic refactor
- introduce new command

wallrj replied:
Done.

neolit123 commented on cmd/kubeadm/app/cmd/alpha/certs_test.go:
```diff
@@ -294,28 +294,30 @@ func TestRenewUsingCSR(t *testing.T) {
+
+	expectedKubeConfigs := []string{
+		"admin",
+		"kubelet",
```
the kubelet.conf is managed by the kubelet (but "kubeadm init" does amend it, at one place/phase). i do not expect the new command to generate CSRs for the kubelet, especially in a case where kubelet client certificates are auto-rotated post TLS bootstrap.

wallrj replied:
Ideally, I'd want to start the apiserver with only an external CA certificate.
Which means that the the control-plane kubelet, which uses a client-certificate based kubeconfig, rather than a bootstrap token based kubeconfig, will also need to have a client certificiate signed by the external CA.

I'm going to test that today, to make sure I've understood this correctly.

neolit123 replied:
normally the TLS bootstrap process would create the CSR and let the KCM sign it.
for kubelet client certs this is done automatically, and i'd assume the new `signerName` feature and external signer support in general will have to manage that.

i don't see the need for creating a CSR / key on disk for this unless the user wants to manually rotate their kubelet client certificate, which conflates with the existing client certificate that the kubelet / KCM normally manage when kubelet client certificate rotation is enabled (enabled by default by kubeadm).

but please, double check my thinking as well.

neolit123 replied:
after our discussion during the kubeadm office hours we decided to leave this for now.

neolit123 commented on cmd/kubeadm/app/cmd/alpha/certs.go:
```diff
@@ -170,36 +169,21 @@ func newCmdGenCSR() *cobra.Command {
 func runGenCSR(config *genCSRConfig) error {
-	certDir := config.kubeadmConfig.CertificatesDir
-	certTree, err := certsphase.GetDefaultCertList().AsMap().CertTree()
+	leafCertificates, err := certsphase.LeafCertificates(certsphase.GetDefaultCertList())
```
non-blocking nit (no need to amend). ideally the refactors should have been before the addition of the new logic.

neolit123 commented on cmd/kubeadm/app/cmd/alpha/certs_test.go:
```diff
@@ -294,28 +294,30 @@ func TestRenewUsingCSR(t *testing.T) {
-apiVersion: kubeadm.k8s.io/v1beta2
-kind: ClusterConfiguration
-certificatesDir: /custom/config/certificates-dir
-kubernetesVersion: v1.19.0
```
e.g. from my previous comment, the PR adds this constant in one commit and then removes it in another.

neolit123 commented on cmd/kubeadm/app/phases/certs/certlist.go:
```diff
@@ -204,6 +207,22 @@ func (c Certificates) AsMap() CertificateMap {
+
+func (c Certificates) visit(visitor certificatesVisitor) error {
+	if visitor == nil {
+		return fmt.Errorf("%w: visitor was nil", errInvalid)
```
`errors.Errorf(....)`

neolit123 commented on cmd/kubeadm/app/phases/certs/certlist.go:
```diff
@@ -397,3 +416,58 @@ func setCommonNameToNodeName() configMutatorsFunc {
+	kubeadmConfig *kubeadmapi.InitConfiguration
+}
+
+func (o *keyAndCSRCreator) create(cert *KubeadmCert) error {
```
please replace usages of fmt.Errorf() in this function with errors.Errorf

neolit123 commented on cmd/kubeadm/app/phases/certs/certlist.go:
```diff
@@ -397,3 +416,58 @@ func setCommonNameToNodeName() configMutatorsFunc {
+	certDir := o.kubeadmConfig.CertificatesDir
+	name := cert.BaseName
+	if pkiutil.CSROrKeyExist(certDir, name) {
+		return fmt.Errorf("%w: key or CSR %s/%s", os.ErrExist, certDir, name)
```
we can avoid os.ErrExit here and the os import.

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -486,3 +358,141 @@ func ValidateKubeconfigsForExternalCA(outDir string, cfg *kubeadmapi.InitConfigu
+// executes the supplied visitor function with each key, value pair.
+func (o KubeConfigs) visit(visitor kubeConfigsVisitor) error {
+	if visitor == nil {
+		return fmt.Errorf("%w: visitor was nil", errInvalid)
```
errors.Errorf

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -486,3 +358,141 @@ func ValidateKubeconfigsForExternalCA(outDir string, cfg *kubeadmapi.InitConfigu
+// GetDefaultKubeConfigs returns all KubeConfigSpecs without CA key or certificate.
+func GetDefaultKubeConfigs(cfg *kubeadmapi.InitConfiguration) (KubeConfigs, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("%w: cfg was nil", errInvalid)
```
errors.Errorf

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -486,3 +358,141 @@ func ValidateKubeconfigsForExternalCA(outDir string, cfg *kubeadmapi.InitConfigu
+		return fmt.Errorf("%w: name was empty", errInvalid)
+	}
+	if spec == nil {
+		return fmt.Errorf("%w: spec was nil", errInvalid)
```
please use errors.Wrap / Errorf

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -486,3 +358,141 @@ func ValidateKubeconfigsForExternalCA(outDir string, cfg *kubeadmapi.InitConfigu
+		return err
+	}
+	if pkiutil.CSROrKeyExist(o.kubeConfigDir, name) {
+		return fmt.Errorf("%w: csr: %s", os.ErrExist, kubeConfigPath)
```
ditto.

neolit123 commented:
@wallrj overall, i have concerns about last minute refactors of existing logic 2 days before feature freeze. especially given i will be the only active reviewer until then.

this PR should have made the following changes:
- deprecate the flags for "init phase certs" (this is done correctly in commit 1 and as discussed not doing that for "renew" for now)
- if needed, apply some refactors to accommodate the addition of the new command
- add the new command
- add unit tests to verify the command main function works on an alpha-grade level, meaning we don't need 100% coverage

instead we ended up with a potentially desired refactor, but on a much larger scale which can be problematic to agree on so close to freeze.

my proposal here would be the following.
- minimize this PR as much as possible (see above)
- create a new PR that will refactor the code and let's merge that once k/k opens for 1.20.

wallrj committed:
Add new helper functions for creating keys, kubeconfig and CSR files

Signed-off-by: Richard Wall <richard.wall@jetstack.io>

wallrj committed:
kubeadm alpha certs generate-csr

* Creates private keys and CSR files for all the control-plane certificates
* Helps with External CA mode of kubeadm

Signed-off-by: Richard Wall <richard.wall@jetstack.io>

wallrj commented:
@neolit123 Thanks for the review, and sorry for all the code churn.

I've rebased and simplified as you suggested, 
* minimised the changes to the existing phases/certs and phases/kubeconfig packages
* moving those changes to a separate refactoring commit
* removed the unit tests for those changes (I'll submit them as a followup PR)

This reduces the diff back to ~500 lines.
I'll join the office hours meeting this afternoon to discuss further.

I totally understand if you prefer to defer these changes for the next release.

neolit123 commented on cmd/kubeadm/app/phases/certs/certlist.go:
```diff
@@ -397,3 +397,68 @@ func setCommonNameToNodeName() configMutatorsFunc {
+func leafCertificates(c Certificates) (Certificates, error) {
+	certTree, err := c.AsMap().CertTree()
+	if err != nil {
+		return nil, errors.WithStack(err)
```
we should just return `err` or errors.Wrap(f) in instead of WithStack (applies to the rest of the diff too)

neolit123 commented on cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:
```diff
@@ -386,3 +357,117 @@ func ValidateKubeconfigsForExternalCA(outDir string, cfg *kubeadmapi.InitConfigu
+	}, nil
+}
+
+var errExist = errors.New("file already exists")
```
we should move all err declarations near the top of the file.

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers.go:
```diff
@@ -290,11 +290,9 @@ func TryLoadKeyFromDisk(pkiPath, name string) (crypto.Signer, error) {
+	csr, err := TryLoadCSRFromDisk(pkiPath, name)
 	if err != nil {
-		return nil, nil, errors.Wrapf(err, "couldn't load the certificate request %s", csrPath)
+		return nil, nil, errors.Wrap(err, "couldn't load csr file")
```
csr -> CSR

neolit123 commented on cmd/kubeadm/app/util/pkiutil/pki_helpers.go:
```diff
@@ -334,6 +332,18 @@ func TryLoadPrivatePublicKeyFromDisk(pkiPath, name string) (*rsa.PrivateKey, *rs
+
+	csr, err := CertificateRequestFromFile(csrPath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "couldn't load the certificate request %s", csrPath)
```
"couldn't load the certificate request" -> "could not load the CSR"

neolit123 reviewed (commented):
@wallrj after the last refactor to limit the number of changes before code freeze this looks good.
once the last few comments are resolved i can add `lgtm`.

the release note:
> `kubeadm alpha certs generate-csr` is a new command  which allows you to generate new private keys and  certificate signing requests for all the control-plane components so that the certificates can be signed by an external CA.

is still missing the deprecation notice about the -csr* flags for `kubeadm init phase certs`.

/approve
/hold for the final change requests

wallrj commented:
@neolit123 Thanks for that latest review. I've made the following changes:
 * Removed all `errors.WithStack` calls. I had misunderstood how the errors package stack traces work. I see now that the stack traces should only be added where you first wrap a non-errors error. There after you should just return the err or use WithMessage to add context messages. Using `errors.Wrap` multiple times causes multiple stack traces to be printed. See https://github.com/pkg/errors/issues/75 and its comments.
 * Replaced the `errInvalid` and `errExists` error types with string constants which I then prepend to the corresponding errors messages when they are created. The previous use of `errors.New` was causing the stacktrace to start at the point where the errors are defined, rather than where they are returned. Perhaps in future I could use stdlib errors package to create those error types and then wrap them with the errors package.
 * Rebased on top of latest master
 * Updated the release notes in the description to include a note about deprecating the --csr-??? flags.

An E2E test has failed, but I assume it's a flake.

I have also tested the new command to verify that the generated CSRs can be signed by an external CA and then used by kubeadm init in External CA mode. It seems to work. Here's the script I'm using: https://github.com/cert-manager/signer-venafi/pull/6 and a screen cast of it running: https://github.com/cert-manager/signer-venafi/blob/68d32e2aa4da21f95d823d4f26201926d18b66cf/docs/demos/kubeadm-bootstrap/README.md#demo-script

wallrj commented:
/retest

neolit123 reviewed (commented):
thanks @wallrj 
/lgtm
/hold cancel

neolit123 commented:
/retest
